File: kotlin/dnd-character/src/main/kotlin/DndCharacter.kt
import kotlin.random.Random

class DndCharacter {
    val strength: Int = ability()
    val dexterity: Int = ability()
    val constitution: Int = ability()
    val intelligence: Int = ability()
    val wisdom: Int = ability()
    val charisma: Int = ability()
    val hitpoints: Int = 10 + modifier(constitution)

    companion object {
        fun ability(): Int = List(4) { Random.nextInt(1, 6.inc()) }.sorted().takeLast(3).sum()

        fun modifier(score: Int) = Math.floorDiv(score - 10, 2)
    }
}

File: elixir/dnd-character/lib/dnd_character.ex
defmodule DndCharacter do
  defstruct ~w[strength dexterity constitution intelligence wisdom charisma hitpoints]a

  def modifier(score), do: Integer.floor_div(score - 10, 2)

  def ability do
    Enum.take_random(1..6, 4)
    |> Enum.sort()
    |> Enum.drop(1)
    |> Enum.sum()
  end

  def character do
    constitution = ability()

    %DndCharacter{
      strength: ability(),
      dexterity: ability(),
      constitution: constitution,
      intelligence: ability(),
      wisdom: ability(),
      charisma: ability(),
      hitpoints: 10 + modifier(constitution)
    }
  end
end
